README.md
# ![Ejtimaa logo](/app/public/images/logo.png) **WebRTC meeting service using [mediasoup](https://mediasoup.org).**
Official website: [Ejtimaa.org](https://Ejtimaa.org)

https://user-images.githubusercontent.com/37835902/152279867-639db9bc-bf78-430f-b96f-d17733527474.mp4

Try it online at [letsmeet.no](https://letsmeet.no)

## Main features

| Feature  | Description |
| ------------- | ------------- |
| **A/V streaming** | Share your microphone and camera + additional video stream  |
| **Video layouts** | Choose between **Democratic** and **Filmstrip** views. More in progress. |
| **Screen sharing** | Share your screen to make some presentation right from your desktop |
| **File sharing** | Share your files with the peers (torrent solution under the hood) |
| **Chat messages**  | Text conversation with other participants |
| **Local Recording**  | Record window/tab/screen content in browser supported formats with room audio and save them (**disabled by default**) |
| **Authentication**  | Supported types: **OIDC**, **SAML**, **local db (text-based)** |


### Internationalization (22 languages) 
<details>
  <summary>Help us with translations:exclamation:</summary>
  
  #### How to contribute?
	
  1. Continue to translate existing [language file](/app/src/intl/translations)
  2. find the _null_  values
  >	"settings.language": null,
  3. replace them based on the _en.json_ file
  >	"settings.language": "Select language",
  4. If your language is not listed, create a new translation _.json_ file..
  >	copy en.json to [_"two letter country code"_.json](https://en.wikipedia.org/wiki/ISO_3166-1_alpha-2) and translate to your languange 
  5. make a Pull Request, or send us a file by [e-mail](mailto:community@lists.Ejtimaa.org)
	
  Thank you in advance!
</details>


### Local Recording
<details>
  <summary>See more</summary>
  
* Local Recording records the browser window video and audio. From the list of media formats that your  browser supports you can select your preferred media format in the settings menu advanced video menu setting.  MediaRecorder makes small chucks of recording and these recorded blob chunks temporary stored in IndexedDB, if IndexedDB implemented in your browser. Otherwise it stores blobs in memory in an array of blobs.
Local Recording creates a local IndexedDB with the name of the starting timestamp (unix timestamp format)  And a storage called chunks. All chunks read in an array and created a final blob that you can download. After blobs array concatenation as a big blob, this big blob saved as file, and finally we delete the temporary local IndexedDB.

* Local recording is **disabled** by default. It could be enabled by setting _localRecordingEnabled_ to true in  (./app/public/config/config.js)

* **WARNING**: Take care that local recording will increase cpu, memory and disk space consumption.
**Enough free disk space has to be provided!!!**
Keep in mind that Browsers don't allow to use all the disk free capacity!
See more info about browsers storage limits:
  * <https://developer.mozilla.org/en-US/docs/Web/API/IndexedDB_API/Browser_storage_limits_and_eviction_criteria#storage_limits>
  * <https://chromium.googlesource.com/chromium/src/+/refs/heads/master/storage/browser/quota/quota_settings.cc#68>

</details>

# Installation 

See here for [Docker](https://github.com/Ejtimaa/Ejtimaa-docker/) or [Ansible](https://github.com/Ejtimaa/Ejtimaa-ansible/) (based on Docker) installation procedures

## Debian & Ubuntu based operating systems (.deb package)

* Prerequisites: Installed NodeJS (v16.x) as described in [Manual installation](#manual-installation-build) section.
* See [Configuration](#configuration) section for client and server configuration details.
* Download from [releases](https://github.com/Ejtimaa/Ejtimaa/releases) assets, or latest job [artifact](https://github.com/Ejtimaa/Ejtimaa/actions?query=workflow%3ADeployer+branch%3Amaster+is%3Asuccess). 

```bash
# Unzip the file
unzip Ejtimaa.zip

# Install the package
sudo apt install Ejtimaa/Ejtimaa.deb

# After package installation, don't forget to edit configuration files.
sudo nano /etc/educonf/client-config.js
sudo nano /etc/educonf/server-config.js
sudo nano /etc/educonf/server-config.yaml

# Finally, start the service by (it's enabled by default)
sudo systemctl start Ejtimaa
```

## Manual installation (build)
Installation example is based on Debian/Ubuntu Linux operating system.
1. Install [NodeJS (v16.x)](https://github.com/nodesource/distributions) and [Yarn ](https://classic.yarnpkg.com/en/docs/install#debian-stable) package manager 
- NodeJS (v16.x) [Debian/Ubuntu](https://github.com/nodesource/distributions#deb)
```bash
# Using Ubuntu
curl -fsSL https://deb.nodesource.com/setup_16.x | sudo -E bash -
sudo apt-get install -y nodejs

# Using Debian, as root
curl -fsSL https://deb.nodesource.com/setup_16.x | bash -
apt-get install -y nodejs
```
- Yarn package manager:
```bash
curl -sS https://dl.yarnpkg.com/debian/pubkey.gpg | sudo apt-key add -
echo "deb https://dl.yarnpkg.com/debian/ stable main" | sudo tee /etc/apt/sources.list.d/yarn.list
sudo apt update && sudo apt install yarn
```
2. Install all required dependencies
```bash
sudo apt update && sudo apt install -y curl git python python3-pip build-essential redis openssl libssl-dev pkg-config
```	
3. Clone Ejtimaa git repository
```bash
git clone https://github.com/Ejtimaa/Ejtimaa.git
cd Ejtimaa
```
(switch to the "develop" branch to get the latest features)
```bash
git checkout develop 
```
### Configuration
**Ejtimaa** will start and run normally with just default settings. If there is no configuration files,  it will automatically detect your host IP address, and listen on port 443 (https). In order to change default values (e.g. certificates), or activate features (e.g. authentication), use appropriate configuration file (see below for details).  

**:warning: Note:** There are separate configuration files for Ejtimaa application and Ejtimaa server:

**Ejtimaa application (app)** for: enabling login, change logo or background, adjust A/V parameters, etc...

Copy [example](/app/public/config/config.example.js) template and edit values (see all available parameters in [./app/public/config/README.md](/app/public/config/README.md))
```bash
cp app/public/config/config.example.js app/public/config/config.js
```

**Ejtimaa server** require **:warning:two** configuration files: **config.js**, and **config.{_json_, _yaml_ or _toml_}** (multiple format supported)

**1. config.js** for setting authentication methods and user roles.

Copy example template and edit values (see additional details in [example](/server/config/config.example.js) file)
```bash
cp server/config/config.example.js server/config/config.js
```

**2. config.{_json_, _yaml_ or _toml_}** for configuring: server port, server certificates, [STUN/TURN](#turn-configuration) configuration, monitoring, etc...  (See below examples of different configuration styles). 

[**:point_right: _config.yaml_**](/server/config/config.example.yaml) example:
```yaml 
    listeningPort: 443
    tls:
        key:  /opt/Ejtimaa/server/certs/privkey.pem
        cert: /opt/Ejtimaa/server/certs/cert.pem
```
[**:point_right: _config.json_**](/server/config/config.example.json) example:
```javascript
    {
        "listeningPort" : "443",
        "tls" : {
            "cert" : "/opt/Ejtimaa/server/certs/cert.pem",
            "key"  : "/opt/Ejtimaa/server/certs/privkey.pem"
        }
    }
```
[**:point_right: _config.toml_**](/server/config/config.example.toml) example:
```toml
    listeningPort = "443"

    [tls]
    cert = "/opt/Ejtimaa/server/certs/cert.pem"
    key = "/opt/Ejtimaa/server/certs/privkey.pem"
```
**:red_circle: IMPORTANT:**  Use **only one** type for second configuration file (`yaml` file format is highly recommended)

Copy **only one** example template file and edit values (see all available parameters in [./server/config/README.md](/server/config/README.md))
```bash
cp server/config/config.example.yaml server/config/config.yaml
  OR!!!
cp server/config/config.example.json server/config/config.json
  OR!!!
cp server/config/config.example.toml server/config/config.toml
```

**:warning: NOTE:** application and server components **has to be rebuild** if configuration parameter is changed ([see build steps](#manual-installation-build)). Rebuild is not necessary for Docker or Debian (.deb) version, just restart container/service.

### Build
**Note:** It is highly recommended to use _yarn_ package manager.

```bash
cd app
yarn && yarn build

cd ../server
yarn && yarn build
```
### Run

**Run on server** (as root or with sudo) 

```bash
# Run the Node.js server application in a terminal:
cd server
sudo yarn start
```

**Run locally** (for development)

* The newest build is always in **develop branch** if you want to make a contribution/pull request use it instead of master branch.

```bash
# run a live build from app folder:
app$ yarn start

# and run server in server folder: 
server$ yarn start
```

Note: To avoid running server as root, redirects privileged ports with firewall rules:
```bash
#adjust ports to your needs

sudo iptables -t nat -A PREROUTING -p tcp --dport 443 -j REDIRECT --to-ports 8443
sudo iptables -t nat -A OUTPUT -p tcp --dport 443 -o lo -j REDIRECT --to-port 8443
sudo iptables -t nat -A PREROUTING -p tcp --dport 3443 -j REDIRECT --to-ports 8443
sudo iptables -t nat -A OUTPUT -p tcp --dport 3443 -o lo -j REDIRECT --to-port 8443

# make it persistent
sudo apt install iptables-persistent
sudo iptables-save > /etc/iptables/rules.v4
sudo ip6tables-save > /etc/iptables/rules.v6
```
* Test your service in a webRTC enabled browser: `https://yourDomainOrIPAdress:3443/roomname`

**Run as a service** (systemd)

```bash
# Stop your locally running server. Copy systemd-service file `Ejtimaa.service` to `/etc/systemd/system/` and check location path settings:
cp Ejtimaa.service /etc/systemd/system/

# modify the install paths, if required
sudo edit /etc/systemd/system/Ejtimaa.service

# Reload systemd configuration and start service:
sudo systemctl daemon-reload
sudo systemctl start Ejtimaa

# If you want to start Ejtimaa at boot time:
sudo systemctl enable Ejtimaa
```

## Ports and firewall
| Port | protocol | description |
| ---- | ----------- | ----------- |
|  443 | tcp | default https webserver and signaling - adjustable in `server/config/config.yaml`) |
| 4443 | tcp | default `yarn start` port for developing with live browser reload, not needed in production environments - adjustable in app/package.json) |
| 40000-49999 | udp, tcp | media ports - adjustable in `server/config/config.yaml` |

## Load balanced installation

To deploy this as a load balanced cluster, have a look at [HAproxy](HAproxy.md).

## Learning management integration

To integrate with an LMS (e.g. Moodle), have a look at [LTI](LTI/LTI.md).

## TURN configuration

If you are part of the GEANT eduGAIN, you can request your turn api key at [https://turn.geant.org/](https://turn.geant.org/)
	
You need an additional [TURN](https://github.com/coturn/coturn)-server for clients located behind restrictive firewalls! 
Add your server and credentials to `server/config/config.yaml`

## Community-driven support
| Type                |                                                |
| -----------         | -----------                                    |
| Open mailing list   | community@lists.Ejtimaa.org                    |
| Subscribe           | lists.Ejtimaa.org/sympa/subscribe/community/   |
| Open archive        | lists.Ejtimaa.org/sympa/arc/community/         |

## Authors

* Håvar Aambø Fosstveit
* Stefan Otto
* Mészáros Mihály
* Roman Drozd
* Rémai Gábor László
* Piotr Pawałowski

This started as a fork of the [work](https://github.com/versatica/mediasoup-demo) done by:

* Iñaki Baz Castillo [[website](https://inakibaz.me)|[github](https://github.com/ibc/)]

## License

MIT License (see `LICENSE.md`)


CHANGELOG.md
# Changelog

## 3.5.2 20-04-2022
* Added  theme
* File Efect 
  1. TopAppBar
  2. toolareaAction
  3. toolarea
  4. Room
## 3.5.1
* Updated Documentation

## 3.5.0

### Added

* Update build steps
* Update documentations
* Major chat refactoring
  * Added chat badge displaying unread messages
  * Merged file sharing + chat
  * Save chat history (both in leave screen and in chat directly)
  * Sorting messages
  * Format messages
  * Combine messages
* Local recording (chrome, firefox, edge)
  * This is still default off in config because of bad UX connected to getUserMedias
* New app configuration system
* New server configuration system
* Add leaving room confirmation dialog
* New audio config with audio profiles
* New translations
* Screen Sharing now supports audio (chrome, edge)
* Add QoS set DSCP bits in IP header according to new option networkPriorities in app config
  Our media traffic should get now higher priority where QoS has been implemented.
  Routers will add higher priority or put media packets to low latency queue.
  If QoS implemented in L2 too, then switches also set the L2 CoS bits according DSCP.
  WIFI AP-s that implements QoS, also based on it could activate the multimedia extension MMS.
  Hopefully it will help and add little bit more priority lower latency in congested networks.
  At the time of this writing it is only implemented in Chrome.

### Improved
* Scalability and performance: 
  * Improvements for handling simulcast and video container size
  * Improved performance for encryption of media (+25% capacity)
  * Improved balancing of peers to workers
* CSS for mobile
* UI improvements (bugs fixes) #922
* Improved button layout
* Improved login dialog
* Added bcrypt for local auth strategy
 

### Fixed 
* Fix auto unmute buttons on separate bar
* Fix browser version comparison
* Safari Screen Share Bug Fix
* Fix session overwrites
* Fix loosing the current language after leaving the room)
* Extra video duplication fix
* Fix loosing the current language after leaving the room)
* Fix notifications (they were displayed always in English)
* Translation updates
* A lot of bugs in auth
* Sanitizations for room names

## 3.4.0

### Added

* Multiparty meeting renamed to Ejtimaa
* Merged room selector- and join- dialog - default keyboard focus on login field
* Localization selectable
* Added aspect ratio 16 : 9  and this is default now, ref #439
* New worker load calculation so router selection is based on that and not random anymore
* New permissions and roles:
  * Propagate userRoles to client state, ref #437
  * Extend userRoles and use the new audio, video permissions, ref #437
  * New permission to modify peer roles
  * Create room actions for giving and taking roles
  * Add functions to client for modifying roles live
  * Ability to give roles to users
  * Add new permission to config
* Add room to userMapping. Example of giving Host if there is no authenticated user
* Admit all peers from lobby when a peer joins with the PROMOTE_PEER permission and activateOnHostJoin is true in config
* Make menus more intuitive on mobile
* Simplify electron screenshare check
* Logo support
* Improve autoMute mic-indicators, Improve audio level scaling
* Clean up participant list
* Add indicator for peers in focus. Ref #360
* TCP enabled by default, prefer UDP
* Ability for Prometheus exporter to listen on localhost
* Make list headers bolder
* Trim displayName inputs. Add random number to Guest displayName.
* Removed facingMode from mobile
* Documentation for prometheus exporter
* Added switching of own video mirror in settings
* Added hiding of own videos
* Request audio and webcam permission at once, when user is requesting media
* Add initial support for local and saml auth
  * Local login form
  * Add bcrypt encrypted passwords for local strategy
  * Add displayName mapping to usermapping
  * Add saml attriute mappings
* Use shared cookieparser for web and websocket
* Update TopBar leave button
* Add joinAudio capability
* Standardize Auth button

### Migration

from last master version:
* Copy paste defaultAudio and centralAudioOtions from config.example.js to config.js 
* Copy paste whole theme from config.example.js to config
* Configure logo in config.js

### Upgrade depencies

* webtorrent from 0.107.17 to 0.108.1 …
* Upgrade React-scripts

### Languages

* Updates translations: hu, tr, no, pl, uk
* Addad translations: Hindi (hi), Russian (ru), Kazakh (ka)

### Bugfixes

* Hopefully fix to silent peer issue, ref #256
* Set timeout for TURN API request, fixes #484
* Possible fix for #582 Crackling sound
* Fix for #444 Settings persistence
* Fix Audio settings from config.js take no effect
* Fix only firefox handles applyConstraints to audio tracks correctly
* Fix (autoMuted) mic too big click area
* Fix userRoles check
* Fix spoltights ignoring maxLastN
* Fix wrong config parameter naming (voiceActivatedUnmute)
* Fix locale checking state
* Fix express error handler
* Add constraint for user facing camera. Update media devices properly
* Cleanup on close
* Fixed null values for peers not yet assigned to a router
* Fix URL sanitizer bug
* Removed audio request modification
* Fix race in spotlights
* Fix piping bug when peers returning to router
* Fix piped router count
* Add comma to prometheus config code
* Removed code that is problematic and also unused
* Remove duplicated callbackURL
* Fix Host buttons layout
* Fix: Special chars are not sanitized in URL
* Fix another roomId bug
* Tidy: replace obj.entries with obj.values to avoid unused key
* Fix close room link

## 3.3

* Add: Rooms now scale across cores
* Add: Permissions and roles. Users can now have different roles (Host, admin etc.) that give different permissions.
* Add: TURN API or fallback TURN server
* Add: Configurable room size limit
* Add: Prometheus monitoring support
* Add: Possible to share several videos (ex: 2 webcams)
* Add: Configurable audio settings (echocancellation etc.)

app/src/config.ts
import convict from 'convict';
import * as convictFormatWithValidator from 'convict-format-with-validator';

convict.addFormat(convictFormatWithValidator.url);

function assert(assertion: Boolean, msg: string)
{
	if (!assertion)
		throw new Error(msg);
}

convict.addFormat({
	name     : 'float',
	coerce   : (v: string) => parseFloat(v),
	validate : (v: number) => assert(Number.isFinite(v), 'must be a number')
});

/**
 * The Ejtimaa configuration schema.
 *
 * Use `yarn gen-config-docs` to re-generate the README.md and the
 * public/config/config.example.js files.
 */
const configSchema = convict({
	loginEnabled :
	{
		doc     : 'If the login is enabled.',
		format  : 'Boolean',
		default : false
	},
	developmentPort :
	{
		doc     : 'The development server listening port.',
		format  : 'port',
		default : 3443
	},
	productionPort :
	{
		doc     : 'The production server listening port.',
		format  : 'port',
		default : 443
	},
	serverHostname :
	{
		doc      : 'If the server component runs on a different host than the app you can specify the host name.',
		format   : 'String',
		default  : '',
		nullable : true
	},

	/**
	 * Supported browsers version in bowser satisfy format.
	 * See more:
	 * https://www.npmjs.com/package/bowser#filtering-browsers
	 * Otherwise you got a unsupported browser page
	 */
	supportedBrowsers :
	{
		doc     : 'Supported browsers version in bowser satisfy format.',
		format  : Object,
		default :
		{
			'windows' : {
				'internet explorer' : '>12',
				'microsoft edge'   	: '>18'
			},
			'microsoft edge'   	           : '>18',
			'safari'                       : '>12',
			'firefox'                      : '>=60',
			'chrome'                       : '>=74',
			'chromium'                     : '>=74',
			'opera'                        : '>=62',
			'samsung internet for android' : '>=11.1.1.52'
		}
	},

	/**
	 * Network priorities 
	 * DSCP bits set by browser according this priority values. 
	 * ("high" means actually: EF for audio, and AF41 for Video in chrome)
	 * https://en.wikipedia.org/wiki/Differentiated_services
	 */
	networkPriorities :
	{
		doc     : 'Network priorities.',
		format  : Object,
		default :
		{
			audio            : 'high',
			mainVideo        : 'high',
			additionalVideos : 'medium',
			screenShare      : 'medium'
		}
	},
	// The aspect ratio of the videos as shown on the screen. 
	// This is changeable in client settings.
	// This value must match one of the defined values in
	// aspectRatios EXACTLY (e.g. 1.333)
	aspectRatio :
	{
		doc : `The aspect ratio of the videos as shown on the screen.
This value must match exactly one of the values defined in aspectRatios.`,
		format  : 'float',
		default : 1.777
	},
	aspectRatios :
	{
		doc     : 'The selectable aspect ratios in the user settings.',
		format  : Array,
		default :
		[
			{
				value : 1.333, // 4 / 3
				label : '4 : 3'
			},
			{
				value : 1.777, // 16 / 9
				label : '16 : 9'
			}
		]
	},
	resolution :
	{
		doc     : 'The default video camera capture resolution.',
		format  : [ 'low', 'medium', 'high', 'veryhigh', 'ultra' ],
		default : 'medium'
	},
	frameRate :
	{
		doc     : 'The default video camera capture framerate.',
		format  : 'nat',
		default : 15
	},
	screenResolution :
	{
		doc     : 'The default screen sharing resolution.',
		format  : [ 'low', 'medium', 'high', 'veryhigh', 'ultra' ],
		default : 'veryhigh'
	},
	screenSharingFrameRate :
	{
		doc     : 'The default screen sharing framerate.',
		format  : 'nat',
		default : 5
	},
	simulcast :
	{
		doc     : 'Enable or disable simulcast for webcam video.',
		format  : 'Boolean',
		default : true
	},
	simulcastSharing :
	{
		doc     : 'Enable or disable simulcast for screen sharing video.',
		format  : 'Boolean',
		default : false
	},
	simulcastProfiles :
	{
		doc     : 'Define different encodings for various resolutions of the video.',
		format  : Object,
		default :
		{
			3840 :
				[
					{ scaleResolutionDownBy: 12, maxBitRate: 150000 },
					{ scaleResolutionDownBy: 6, maxBitRate: 500000 },
					{ scaleResolutionDownBy: 1, maxBitRate: 10000000 }
				],
			1920 :
				[
					{ scaleResolutionDownBy: 6, maxBitRate: 150000 },
					{ scaleResolutionDownBy: 3, maxBitRate: 500000 },
					{ scaleResolutionDownBy: 1, maxBitRate: 3500000 }
				],
			1280 :
				[
					{ scaleResolutionDownBy: 4, maxBitRate: 150000 },
					{ scaleResolutionDownBy: 2, maxBitRate: 500000 },
					{ scaleResolutionDownBy: 1, maxBitRate: 1200000 }
				],
			640 :
				[
					{ scaleResolutionDownBy: 2, maxBitRate: 150000 },
					{ scaleResolutionDownBy: 1, maxBitRate: 500000 }
				],
			320 :
				[
					{ scaleResolutionDownBy: 1, maxBitRate: 150000 }
				]
		}
	},

	// The adaptive spatial layer selection scaling factor (in the range [0.5, 1.0])
	// example: 
	// with level width=640px, the minimum width required to trigger the
	// level change will be: 640 * 0.75 = 480px
	adaptiveScalingFactor :
	{
		doc     : 'The adaptive spatial layer selection scaling factor in the range [0.5, 1.0].',
		format  : (value: number) => value >= 0.5 && value <= 1.0,
		default : 0.75
	},

	localRecordingEnabled :
	{
		doc     : 'If set to true Local Recording feature will be enabled.',
		format  : 'Boolean',
		default : true
	},

	/**
	 * White listing browsers that support audio output device selection.
	 * It is not yet fully implemented in Firefox.
	 * See: https://bugzilla.mozilla.org/show_bug.cgi?id=1498512
	 */
	audioOutputSupportedBrowsers :
	{
		doc     : 'White listing browsers that support audio output device selection.',
		format  : Array,
		default : [
			'chrome',
			'opera',
			'firefox'
		]
	},
	requestTimeout :
	{
		doc     : 'The Socket.io request timeout.',
		format  : 'nat',
		default : 20000
	},
	requestRetries :
	{
		doc     : 'The Socket.io request maximum retries.',
		format  : 'nat',
		default : 3
	},
	transportOptions :
	{
		doc     : 'The Mediasoup transport options.',
		format  : Object,
		default : {
			tcp : true
		}
	},
	// audio options
	autoGainControl :
	{
		doc     : 'Auto gain control enabled.',
		format  : 'Boolean',
		default : true
	},
	echoCancellation :
	{
		doc     : 'Echo cancellation enabled.',
		format  : 'Boolean',
		default : true
	},
	noiseSuppression :
	{
		doc     : 'Noise suppression enabled.',
		format  : 'Boolean',
		default : true
	},
	voiceActivatedUnmute :
	{
		doc     : 'Automatically unmute speaking above noiseThreshold.',
		format  : 'Boolean',
		default : false
	},
	noiseThreshold :
	{
		doc     : 'This is only for voiceActivatedUnmute and audio-indicator.',
		format  : 'int',
		default : -60
	},
	sampleRate :
	{
		doc     : 'The audio sample rate.',
		format  : [ 8000, 16000, 24000, 44100, 48000 ],
		default : 48000
	},
	channelCount :
	{
		doc     : 'The audio channels count.',
		format  : [ 1, 2 ],
		default : 1
	},
	sampleSize :
	{
		doc     : 'The audio sample size count.',
		format  : [ 8, 16, 24, 32 ],
		default : 16
	},
	opusStereo :
	{
		doc     : 'If OPUS FEC stereo be enabled.',
		format  : 'Boolean',
		default : false
	},
	opusDtx :
	{
		doc     : 'If OPUS DTX should be enabled.',
		format  : 'Boolean',
		default : true
	},
	opusFec :
	{
		doc     : 'If OPUS FEC should be enabled.',
		format  : 'Boolean',
		default : true
	},
	opusPtime :
	{
		doc     : 'The OPUS packet time.',
		format  : [ 3, 5, 10, 20, 30, 40, 50, 60 ],
		default : 20
	},
	opusMaxPlaybackRate :
	{
		doc     : 'The OPUS playback rate.',
		format  : [ 8000, 16000, 24000, 44100, 48000 ],
		default : 48000
	},
	// audio presets profiles
	audioPreset :
	{
		doc     : 'The audio preset',
		format  : 'String',
		default : 'conference'
	},
	audioPresets :
	{
		doc     : 'The audio presets.',
		format  : Object,
		default :
		{
			conference :
			{
				name                 : 'Conference audio',
				autoGainControl      : true, // default : true
				echoCancellation     : true, // default : true 
				noiseSuppression     : true, // default : true 
				// Automatically unmute speaking above noiseThreshold
				voiceActivatedUnmute : false, // default : false 
				// This is only for voiceActivatedUnmute and audio-indicator
				noiseThreshold       : -60, // default -60
				// will not eat that much bandwidth thanks to opus
				sampleRate           : 48000, // default : 48000 and don't go higher
				// usually mics are mono so this saves bandwidth
				channelCount         : 1, // default : 1
				sampleSize           : 16, // default : 16
				// usually mics are mono so this saves bandwidth
				opusStereo           : false, // default : false
				opusDtx              : true, // default : true / will save bandwidth 
				opusFec              : true, // default : true / forward error correction
				opusPtime            : 20, // minimum packet time (10, 20, 40, 60)
				opusMaxPlaybackRate  : 48000 // default : 48000 and don't go higher
			},
			hifi :
			{
				name                 : 'HiFi streaming',
				autoGainControl      : false, // default : true
				echoCancellation     : false, // default : true 
				noiseSuppression     : false, // default : true 
				// Automatically unmute speaking above noiseThreshold
				voiceActivatedUnmute : false, // default : false 
				// This is only for voiceActivatedUnmute and audio-indicator
				noiseThreshold       : -60, // default -60
				// will not eat that much bandwidth thanks to opus
				sampleRate           : 48000, // default : 48000 and don't go higher
				// usually mics are mono so this saves bandwidth
				channelCount         : 2, // default : 1
				sampleSize           : 16, // default : 16
				// usually mics are mono so this saves bandwidth
				opusStereo           : true, // default : false
				opusDtx              : false, // default : true / will save bandwidth 
				opusFec              : true, // default : true / forward error correction
				opusPtime            : 60, // minimum packet time (10, 20, 40, 60)
				opusMaxPlaybackRate  : 48000 // default : 48000 and don't go higher
			}
		}
	},

	autoMuteThreshold :
	{
		doc : `It sets the maximum number of participants in one room that can join unmuted.
The next participant will join automatically muted.
Set it to 0 to auto mute all.
Set it to negative (-1) to never automatically auto mute but use it with caution, 
full mesh audio strongly decrease room capacity!`,
		format  : 'nat',
		default : 4
	},

	background :
	{
		doc      : 'The page background image URL',
		format   : 'String',
		default  : 'images/background.jpg',
		nullable : true
	},

	defaultLayout :
	{
		doc     : 'The default layout.',
		format  : [ 'democratic', 'filmstrip' ],
		default : 'democratic'
	},

	buttonControlBar :
	{
		doc     : 'If true, the media control buttons will be shown in separate control bar, not in the ME container.',
		format  : 'Boolean',
		default : false
	},

	drawerOverlayed :
	{
		doc : `If false, will push videos away to make room for side drawer.
If true, will overlay side drawer over videos.`,
		format  : 'Boolean',
		default : true
	},

	notificationPosition :
	{
		doc     : 'The position of the notifications.',
		format  : [ 'left', 'right' ],
		default : 'right'
	},

	notificationSounds :
	{
		doc : `It sets the notifications sounds.
Valid keys are: 'parkedPeer', 'parkedPeers', 'raisedHand', 
'chatMessage', 'sendFile', 'newPeer' and 'default'.
Not defining a key is equivalent to using the default notification sound.
Setting 'play' to null disables the sound notification.		
`,
		format  : Object,
		default :
		{
			chatMessage : {
				play : '/sounds/notify-chat.mp3'
			},
			raisedHand : {
				play : '/sounds/notify-hand.mp3'
			},
			default : {
				delay : 5000, // minimum delay between alert sounds [ms]
				play  : '/sounds/notify.mp3'
			}
		}
	},

	hideTimeout :
	{
		doc     : 'Timeout for auto hiding the topbar and the buttons control bar.',
		format  : 'int',
		default : 3000
	},
	lastN :
	{
		doc     : 'The maximum number of participants that will be visible in as speaker.',
		format  : 'nat',
		default : 4
	},
	mobileLastN :
	{
		doc     : 'The maximum number of participants that will be visible in as speaker for mobile users.',
		format  : 'nat',
		default : 1
	},
	maxLastN :
	{
		doc     : 'The highest number of lastN the user can select manually in the user interface.',
		format  : 'nat',
		default : 1000
	},
	lockLastN :
	{
		doc     : 'If true, the users can not change the number of visible speakers.',
		format  : 'Boolean',
		default : false
	},

	logo :
	{
		doc      : 'If not null, it shows the logo loaded from the specified URL, otherwise it shows the title.',
		format   : 'url',
		default  : 'images/logo.png',
		nullable : true
	},
	admit :
	{
		doc      : 'If not null, it shows the admit loaded from the specified URL, otherwise it shows the title.',
		format   : 'url',
		default  : 'images/admit.png',
		nullable : true
	},
	title :
	{
		doc     : 'The title to show if the logo is not specified.',
		format  : 'String',
		default : 'Ejtimaa'
	},
	supportUrl :
	{
		doc      : 'The service & Support URL; if `null`, it will be not displayed on the about dialogs.',
		format   : 'url',
		default  : 'https://support.example.com',
		nullable : true
	},
	privacyUrl :
	{
		doc      : 'The privacy and data protection external URL or local HTML path.',
		format   : 'String',
		default  : 'privacy/privacy.html',
		nullable : true
	},

	theme :
	{
		doc     : 'UI theme elements colors.',
		format  : Object,
		default :
		{
			palette :
			{
				primary :
				{
					main : '#313131'
				}
			},
			overrides :
			{
				MuiAppBar :
				{
					colorPrimary :
					{
						backgroundColor : '#313131'
					}
				},

				MuiButton :
				{
					containedPrimary :
					{
						backgroundColor : 'rgba(0, 0, 0, 0.87)',
						'&:hover'       :
						{
							backgroundColor : 'rgba(0, 0, 0, 0.87)'
						}
					},
					containedSecondary :
					{
						backgroundColor : 'red',
						'&:hover'       :
						{
							backgroundColor : 'red'
						}
					}

				},

				/*
				MuiIconButton :
				{
					colorPrimary :
					{
						backgroundColor : 'rgba(0, 0, 0, 0.87)',
						'&:hover'	   :
						{
							backgroundColor : 'rgba(0, 0, 0, 0.87)'
						}
					},
					colorSecondary :
					{
						backgroundColor : 'red',
						'&:hover'	   :
						{
							backgroundColor : 'red'
						}
					}

				},
				*/

				MuiFab :
				{
					primary :
					{
						backgroundColor : '#e89b3f',
						'&:hover'       :
						{
							backgroundColor : '#e89b3f'
						},
						'&:disabled' : {
							color           : '#999898',
							backgroundColor : '#323131'
						}

					},

					secondary :
					{
						backgroundColor : 'red',
						'&:hover'       :
						{
							backgroundColor : 'red'
						},
						'&:disabled' : {
							color           : '#999898',
							backgroundColor : '#323131'
						}

					}

				},

				MuiBadge :
				{
					colorPrimary :
					{
						backgroundColor : 'rgba(0, 0, 0, 0.87)',
						'&:hover'       :
						{
							backgroundColor : '#e89b3f'
						}
					}
				}
			},
			typography :
			{
				useNextVariants : true
			}
		}
	}
});

function formatDocs()
{
	function _formatDocs(docs: any, property: string | null, schema: any)
	{
		if (schema._cvtProperties)
		{
			Object.entries(schema._cvtProperties).forEach(([ name, value ]) =>
			{
				_formatDocs(docs, `${property ? `${property}.` : ''}${name}`, value);
			});

			return docs;
		}
		else if (property)
		{
			docs[property] =
			{
				doc     : schema.doc,
				format  : JSON.stringify(schema.format, null, 2),
				default : JSON.stringify(schema.default, null, 2)
			};
		}

		return docs;
	}

	return _formatDocs({}, null, configSchema.getSchema());
}

function formatJson(data: string)
{
	return data ? `\`${data.replace(/\n/g, '')}\`` : '';
}

function dumpDocsMarkdown()
{
	let data = `# ![Ejtimaa logo](/app/public/images/logo.png) App Configuration properties list:

| Name | Description | Format | Default value |
| :--- | :---------- | :----- | :------------ |
`;

	Object.entries(formatDocs()).forEach((entry: [string, any]) =>
	{
		const [ name, value ] = entry;

		data += `| ${name} | ${value.doc.replace(/\n/g, ' ')} | ${formatJson(value.format)} | \`${formatJson(value.default)}\` |\n`;
	});

	data += `

---

*Document generated with:* \`yarn gen-config-docs\` *from:* [config.ts](src/config.ts).
`;

	return data;
}

function dumpExampleConfigJs()
{
	let data = `/**
 * Ejtimaa App Configuration
 *
 * The configuration documentation is available also:
 * - in the app/README.md file in the source tree
 * - visiting the /?config=true page in a running instance
 */

// eslint-disable-next-line
var config = {
`;

	Object.entries(formatDocs()).forEach((entry: [string, any]) =>
	{
		// eslint-disable-next-line
		let [ name, value ] = entry;

		if (name.includes('.'))
			name = `'${name}'`;

		data += `\n\t// ${value.doc.replace(/\n/g, '\n\t// ')}
\t${name} : ${value.default},
`;
	});

	data += `};

// Generated with: \`yarn gen-config-docs\` from app/src/config.ts
`;

	return data;
}

// run the docs generator
if (typeof window === 'undefined')
{
	import('fs').then((fs) =>
	{
		fs.writeFileSync('public/config/README.md', dumpDocsMarkdown());
		fs.writeFileSync('public/config/config.example.js', dumpExampleConfigJs());
	});
}

//
let config: any = {};
let configError = '';

// Load config from window object
if (typeof window !== 'undefined' && (window as any).config !== undefined)
{
	configSchema.load((window as any).config);
}

// Perform validation
try
{
	configSchema.validate({ allowed: 'strict' });
	config = configSchema.getProperties();
}
catch (error: any)
{
	configError = error.message;
}

// Override the window config with the validated properties.
if (typeof window !== 'undefined')
{
	(window as any)['config'] = config;
}

export {
	configSchema,
	config,
	configError,
	formatDocs
};
